library/src/blas3/trtri_device.h

```diff
@@ -165,4 +165,212 @@ __device__ void trtri_device(rocblas_fill uplo,
+{
+
+    // quick return
+    if(n <= 0)
+        return;
+
+    int tx = hipThreadIdx_x;
+
+    __shared__ T diag1[IB * IB];
+    __shared__ T diag2[IB * IB];
+    __shared__ T sA[IB * IB];
+    __shared__ T temp[IB * IB];
+
+    T* diagP       = tx < n ? diag1 : (tx < 2 * n ? diag2 : sA);
+    int Aoffset    = tx < n ? 0 : n * lda + n;
+    int AInvoffset = tx < n ? 0 : n * ldinvA + n;
+    int index      = tx < n ? tx : (tx < 2 * n ? tx - n : tx - 2 * n);
+    int r          = tx % n;
+    int c          = tx / n;
+
+    // read matrix A into shared memory, only need to read lower part
+    // its inverse will overwrite the shared memory
+
+    if(tx < 2 * n)
+    {
+        if(uplo == rocblas_fill_lower)
+        {
+            for(int i = 0; i < n; i++)
+            {
+                diagP[index + i * n] = i <= index ? A[Aoffset + index + i * lda] : 0.0f;
+            }
+        }
+        else
+        { // transpose A in sA if upper
+            for(int i = n - 1; i >= 0; i--)
+            {
+                diagP[(n - 1 - index) + (n - 1 - i) * n] =
+                    i >= index ? A[Aoffset + index + i * lda] : 0.0f;
+            }
+        }
+    }
+    else if(tx < n * 3)
+    {
+        if(uplo == rocblas_fill_lower)
+        {
+            for(int i = 0; i < n; i++)
+            {
+                diagP[index + i * n] = A[n + index + i * lda];
+            }
+        }
+        else
+        { // transpose A in diag1 if upper
+            for(int i = n - 1; i >= 0; i--)
+            {
+                diagP[index + i * n] = A[n * lda + index + i * lda];
+            }
+        }
+    }
+
+    __syncthreads(); // if IB < 64, this synch can be avoided - IB = 16 here
+
+    // invert the diagonal element
+    if(tx < 2 * n)
+    {
+        // compute only diagonal element
+        if(diag == rocblas_diagonal_unit)
+        {
+            diagP[index + index * n] = 1.0;
+        }
+        else
+        { // inverse the diagonal
+            if(diagP[index + index * n] == 0.0)
+            {                                   // notice this does not apply for complex
+                diagP[index + index * n] = 1.0; // means the matrix is singular
+            }
+            else
+            {
+                diagP[index + index * n] = 1.0 / diagP[index + index * n];
+            }
+        }
+    }
+
+    __syncthreads(); // if IB < 64, this synch can be avoided on AMD Fiji
+
+    // solve the inverse of A column by column, each inverse(A)' column will overwrite diag1'column
+    // which store A
+    // this operation is safe
+    if(tx < 2 * n)
+    {
+        for(int col = 0; col < n; col++)
+        {
+
+            T reg = 0;
+            // use the diagonal one to update current column
+            if(index > col)
+            {
+                reg += diagP[index + col * n] * diagP[col + col * n];
+            }
+
+            // __syncthreads(); // if IB < 64, this synch can be avoided on AMD Fiji
```

Comment:
Please advise if these syncthreads can be removed safely for required architectures.

Reply:
Best to leave the __syncthreads(), unless you are seeing a big perf impact.
Its presence depends on the algorithm; are only 64-threads in the block collaborating and need to finish before proceeding, or is all 256 threads? If it is 64 only, it is possible to remove

Reply:
That part of the code is in a branch, only 32 of the 256 threads are collaborating; if anything half the wavefront won't hit the barrier and will introduce more issues if left in. There is a sync after the branch which should keep everything consistent afterwards.